sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,465 @@
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withRowMapper(RowMapper rowMapper) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withUsername(String username) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withPassword(String password) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public void validate() {
+        Preconditions.checkNotNull(dataSource, "dataSource");
+        Preconditions.checkNotNull(query, "query");
+      }
+
+      public void populateDisplayData(DisplayData.Builder builder) {
+        builder.add(DisplayData.item("dataSource", dataSource.getClass().getName()));
+        builder.add(DisplayData.item("query", query));
+        builder.addIfNotNull(DisplayData.item("username", username));
+      }
+
+      public DataSource getDataSource() {
+        return dataSource;
+      }
+
+      public String getQuery() {
+        return query;
+      }
+
+      @Nullable
+      public String getUsername() {
+        return username;
+      }
+
+      @Nullable
+      public String getPassword() {
+        return password;
+      }
+    }
+
+    /**
+     * A {@link DoFn} executing the SQL query to read from the database.
+     */
+    public static class ReadFn<T> extends DoFn<JdbcOptions, T> {
+
+      private final RowMapper<T> rowMapper;
+
+      private ReadFn(RowMapper<T> rowMapper) {
+        this.rowMapper = rowMapper;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        JdbcOptions options = context.element();
+
+        try (Connection connection = (options.getUsername() != null)
+            ? options.getDataSource().getConnection(options.getUsername(), options.getPassword())
+            : options.getDataSource().getConnection()) {
+
+          try (PreparedStatement statement = connection.prepareStatement(options.getQuery())) {
+            try (ResultSet resultSet = statement.executeQuery()) {
+              while (resultSet.next()) {
+                T record = rowMapper.mapRow(resultSet);
+                  context.output(record);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * An interface used by the JdbcIO Write to set the parameters of the {@link PreparedStatement}
+   * used to setParameters into the database.
+   */
+  public interface PreparedStatementSetter<T> extends Serializable {
+    void setParameters(T element, PreparedStatement preparedStatement) throws Exception;
+  }
+
+  /**
+   * A {@link PTransform} to write to a JDBC datasource.
+   */
+  public static class Write<T> extends PTransform<PCollection<T>, PDone> {
+
+    public Write<T> withDataSource(DataSource dataSource) {
+      return new Write<>(writer.withDataSource(dataSource));
```

Comment:
This duplication of builder functions between Write and WriteFn seems unnecessary, as well as the duplication between these options and JdbcOptions.

I recommend renaming JdbcOptions into ConnectionParameters and reusing it between Write and Read transforms. Perhaps even more radically, get rid of username and password parameters completely - just create a serializable interface ConnectionFactory { Connection getConnection() } and let people do whatever they want in there.